src/ast/codegen_llvm.cpp

```diff
@@ -661,87 +663,92 @@ void CodegenLLVM::visit(Call &call)
-     * special asynchronous action. The ID maps to print(). The top argument is either
-     * a value for truncation, or 0 for everything. The div argument divides the output values
-     * by this (eg: for use in nanosecond -> millisecond conversions).
-     * TODO: consider stashing top & div in a printf_args_ like struct, so we don't need to pass
-     * them here via the perfdata output (which is a little more wasteful than need be: I'm using
-     * uint64_t's to avoid "misaligned stack access off" errors when juggling uint32_t's).
-     */
+    auto elements = AsyncEvent::Print().asLLVMType(&b_);
+    StructType *print_struct = b_.GetStructType("print_t", elements, true);
+
     auto &arg = *call.vargs->at(0);
     auto &map = static_cast<Map&>(arg);
-    Constant *const_str = ConstantDataArray::getString(module_->getContext(), map.ident, true);
-    AllocaInst *str_buf = b_.CreateAllocaBPF(ArrayType::get(b_.getInt8Ty(), map.ident.length() + 1), "str");
-    b_.CREATE_MEMSET(str_buf, b_.getInt8(0), map.ident.length() + 1, 1);
-    b_.CreateStore(const_str, str_buf);
-    ArrayType *perfdata_type = ArrayType::get(b_.getInt8Ty(), sizeof(uint64_t) + 2 * sizeof(uint64_t) + map.ident.length() + 1);
-    AllocaInst *perfdata = b_.CreateAllocaBPF(perfdata_type, "perfdata");
+
+    AllocaInst *buf = b_.CreateAllocaBPF(print_struct, "print_" + map.ident);
 
     // store asyncactionid:
-    b_.CreateStore(b_.getInt64(asyncactionint(AsyncAction::print)), perfdata);
+    b_.CreateStore(b_.getInt64(asyncactionint(AsyncAction::print)),
+                   b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(0) }));
 
-    // store top:
-    if (call.vargs->size() > 1)
-    {
-      Integer &top_arg = static_cast<Integer&>(*call.vargs->at(1));
-      Value *top;
-      top_arg.accept(*this);
-      top = expr_;
-      b_.CreateStore(top, b_.CreateGEP(perfdata, {b_.getInt32(0), b_.getInt32(sizeof(uint64_t))}));
-    }
-    else
-      b_.CreateStore(b_.getInt64(0), b_.CreateGEP(perfdata, {b_.getInt64(0), b_.getInt64(sizeof(uint64_t))}));
+    auto id = bpftrace_.maps_[map.ident]->id;
+    auto *ident_ptr = b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(1) });
+    b_.CreateStore(b_.GetIntSameSize(id, elements.at(1)), ident_ptr);
 
-    // store top:
-    if (call.vargs->size() > 2)
+    // top, div
+    for (size_t i = 1; i < 3; i++)
     {
-      Integer &div_arg = static_cast<Integer&>(*call.vargs->at(2));
-      Value *div;
-      div_arg.accept(*this);
-      div = expr_;
-      b_.CreateStore(div, b_.CreateGEP(perfdata, {b_.getInt64(0), b_.getInt64(sizeof(uint64_t) + sizeof(uint64_t))}));
+      llvm::Value *val;
+      if (call.vargs->size() > i)
+      {
+        call.vargs->at(i)->accept(*this);
+        val = b_.CreateIntCast(expr_, elements.at(i), false);
+      }
+      else
+        val = b_.GetIntSameSize(0, elements.at(i));
+
+      b_.CreateStore(val,
+                     b_.CreateGEP(buf, { b_.getInt64(0), b_.getInt32(i + 1) }));
```

Comment:
Might be a personal preference, but this was a little confusing for me to read. Might be more clear if there was 2 loops with something like:

```
size_t i;
for (i = 1; i < call.vargs.size(); i++)
{
  <accept real value and use it>
}

// Fill in 0 for remaining
for (; i < 3; i++)
{
  <use 0>
}
```